[FLINK-9374] [kinesis] Enable FlinkKinesisProducer Backpressuring

fmthoma opened the pull request:
## What is the purpose of the change

The `FlinkKinesisProducer` just accepts records and forwards it to a `KinesisProducer` from the Amazon Kinesis Producer Library (KPL). The KPL internally holds an unbounded queue of records that have not yet been sent.

Since Kinesis is rate-limited to 1MB per second per shard, this queue may grow indefinitely if Flink sends records faster than the KPL can forward them to Kinesis.

One way to circumvent this problem is to set a record TTL, so that queued records are dropped after a certain amount of time, but this will lead to data loss under high loads.

Currently the only time the queue is flushed is during checkpointing: `FlinkKinesisProducer` consumes records at arbitrary rate, either until a checkpoint is reached (and will wait until the queue is flushed), or until out-of-memory, whichever is reached first. (This gets worse due to the fact that the Java KPL is only a thin wrapper around a C++ process, so it is not even the Java process that runs out of memory, but the C++ process.)

My proposed solution is to add a config option `queueLimit` to set a maximum number of records that may be waiting in the KPL queue. If this limit is reached, the `FlinkKinesisProducer` should trigger a `flush()` and wait (blocking) until the queue length is below the limit again. This automatically leads to backpressuring, since the `FlinkKinesisProducer` cannot accept records while waiting. For compatibility, `queueLimit` is set to `Integer.MAX_VALUE` by default, so the behavior is unchanged unless a client explicitly sets the value. Setting a »sane« default value is not possible unfortunately, since sensible values for the limit depend on the record size (the limit should be chosen so that about 10–100MB of records per shard are accumulated before flushing, otherwise the maximum Kinesis throughput may not be reached).

## Brief change log

* Add a `queueLimit` setting to `FlinkKinesisProducer` to limit the number of in-flight records in the Kinesis Producer Library, and enable backpressuring if the limit is exceeded

## Verifying this change

This change added tests and can be verified as follows:

* Added unit test
* Manually verified the change by running a job that produces to a 2-shard Kinesis stream. The input rate is limited by Kinesis (verified that the Kinesis stream is indeed at maximum capacity).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes, but backwards compatible (option was added)
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): don't know
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: don't know
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +342,24 @@ private void checkAndPropagateAsyncError() throws Exception {
+	 */
+	private void checkQueueLimit() {
+		while (producer.getOutstandingRecordsCount() >= queueLimit) {
+			producer.flush();
```
Do we have to do a flush here? Shouldn't the KPL child process process the user records in the background without an explicit flush call?

If so, perhaps a more graceful solution here is to wait on a local object, and notify it to wake up in the asynchronous producer write call backs. After being notified, we check the `getOutstandingRecordsCount` agains the queueLimit, and either wait more or escape the loop.

What do you think?

bowenli86 replied:
Iooks like [KinesisProducer](https://github.com/awslabs/amazon-kinesis-producer/blob/master/java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java) doesn't have a way to get child process's callback. Or maybe I misunderstood your proposal, Gordon?

tzulitai replied:
When we add a record to the producer queue via `producer.addUserRecord(...)`, we get a callback. We can use that callback to notify the blocking operation in `checkQueueLimit`.

fmthoma replied:
So I'd suggest to add `producer.notifyAll()` to both `onSuccess()` and `onFailure()` in the callback, and replace the `Thread.sleep(500)` by `producer.wait(500)`. This way we re-check with every record sent out, or at most after 0.5 seconds.

fmthoma replied:
@tzulitai @bowenli86 I've given this some more thought. `wait()`/`notify()` requires a `synchronized` block. So if we just notify some lock in the callback, this would lead to synchronization overhead. We'd have to recognize a transition from »queue size > queue limit« to »queue size <= queue limit« and only synchronize then, which adds a lot of complexity.

On the other hand: Kinesis accepts up to 1MB per second per shard. The queue limit should be chosen so that some data can be accumulated still before sending, i.e. more than a second of data (more than 1MB per shard). If the queue limit is chosen adequately, then the `Thread.sleep(500)` does not harm, as the queued records take more than one second to flush anyway. If the queue limit is chosen too low, then sleeping half a second may be too long, but we would not reach maximum throughput anyway because of the limitation on the number of `Put` requests.

I think it's not worth the additional complexity.

bowenli86 commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +342,24 @@ private void checkAndPropagateAsyncError() throws Exception {
+	 * We don't want to flush _all_ records at this point since that would
+	 * break record aggregation.
+	 */
+	private void checkQueueLimit() {
```
probably rename it to something different, e.g  `enforceQueueLimit()`? because it clearly does things more than just 'check'

fmthoma replied:
I agree.

bowenli86 commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +342,24 @@ private void checkAndPropagateAsyncError() throws Exception {
+			try {
+				Thread.sleep(500);
+			} catch (InterruptedException e) {
+				LOG.warn("Flushing was interrupted.");
```
you can remove this two lines, they don't provide much value. After removal, it will be almost exactly how `KinesisProducer#flushSync` works

```
// KinesisProducer.java
@Override
    public void flushSync() {
        while (getOutstandingRecordsCount() > 0) {
            flush();
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) { }
        }
    }
```

fmthoma replied:
I don't think so, `flushSync()` will just swallow the interrupt and block again until the queue is empty. `checkQueueLimit()` OTOH aborts immediately on the first interrupt. So there is a difference, although we could of course discuss which one makes more sense.

bowenli86 commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +342,24 @@ private void checkAndPropagateAsyncError() throws Exception {
+	 * break record aggregation.
+	 */
+	private void checkQueueLimit() {
+		while (producer.getOutstandingRecordsCount() >= queueLimit) {
```
A more important thing I would count and log here is how many times it has already tried to flush within a single call of `enforceQueueLimit()`. We can set a threshold, say 10 times, and then log a message saying that KPL is leading to backpressure

fmthoma replied:
You mean, log a message if we have checked more than 10 times (5 seconds) for one record? That makes sense. But we shouldn't log every time we reach the threshold, that would lead to log spam.

bowenli86 commented:
@tzulitai adding docs to educate users on tuning KPL performance would be good. I has quite some experience on it (as you may have know :)  Ping me if you start working on it before I do, and I'll be glad to help contribute

fmthoma commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -218,6 +232,8 @@ public void invoke(OUT value, Context context) throws Exception {
 
+		checkAndPropagateAsyncError();
+		checkQueueLimit();
 		checkAndPropagateAsyncError();
```
`snapshotState()` also checks twice explicitly, and I think it makes sense to have the two checks on the same level. But I won't insist on that, if you prefer having it more implicitly.

fmthoma commented:
@tzulitai I agree on adding additional docs, where do you suggest I should put them? In the Javadoc on `setQueueLimit()`?

My current suggestion is to look at the size of your individual records, and choose the queue limit so that about 10MB per shard are aggregated. 1MB would be too small (since the KPL aggregates the user records to 1MB batches). But I'll run some more performance tests, in particular also with the `wait()`/`notify()` change you suggested above.

tzulitai commented:
@fmthoma I think this might benefit from an actual documentation, not only Javadocs.

fmthoma commented:
@tzulitai I believe the right location is `docs/dev/connectors/kinesis.md`? I'll add some docs there.

tzulitai commented:
@fmthoma yes, that would be great.

fmthoma commented:
@tzulitai @bowenli86 I've made some more changes while investigating awslabs/amazon-kinesis-producer#183:

* I've followed your suggestion and used `wait()` instead of `Thread.sleep()`, see `TimeoutLatch`. This allows much smaller queue sizes.
* The timeout of 500ms is too high, i've lowered it to 100ms.
* I've added two metrics: A `Gauge` for the `outstandingRecordsCount`, and a `Counter` for the backpressure cycles (i.e. the number of times the check `outstandingRecordsCount <= queueLimit` fails).

I updated the documentation, the recommended queue limit is now much lower.

bowenli86 commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/TimeoutLatch.java:
```diff
@@ -0,0 +1,41 @@
+	public void trigger() {
+		if (waiting) {
+			synchronized (lock) {
+				waiting = false;
```
needs another `if (waiting)` here inside the synchronized block, to ensure no one chimes in between line 34 and 35

fmthoma replied:
Why? I don't think a double-check lock is necessary here: There is no harm in setting a variable to `false` that is already `false`, and neither in `notify`ing a lock for which nobody is `wait`ing. But sure, it wouldn't harm, either. Do you insist?

tzulitai replied:
I agree with @fmthoma here.

bowenli86 commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -180,9 +204,16 @@ public void open(Configuration parameters) throws Exception {
 
 		producer = getKinesisProducer(producerConfig);
+
+		final MetricGroup kinesisMectricGroup = getRuntimeContext().getMetricGroup().addGroup("kinesisProducer");
```
minor: better to make these three strings constant (static final String) for easier maintenance.

fmthoma replied:
Sure.

gliu6 commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +366,29 @@ private void checkAndPropagateAsyncError() throws Exception {
+	 * We don't want to flush _all_ records at this point since that would
+	 * break record aggregation.
+	 */
+	private void enforceQueueLimit() {
```
I wonder whether we could adjust the queue limit dynamically. 
you mentioned that `queue limit = (number of shards * queue size per shard) / record size`.
except record size, all others are relatively easy to set. For me, I don't really know the record size until the application starts. Also, what is the record size varies over time?
So how about add a queueLimit supplier function here to allow user to supply how the queueLimit is calculated dynamically?

gliu6 replied:
Btw, I am not requesting any changes, pr looks good to me for it defined purpose.

Just wonder how to config easily. Now I think about this a bit more. Will it be better if we expose `queue size` to the user instead  of `queue limit (number)`, thus, Inside of the FKP class, define an integer recordSize, and inside of the invoke function, do a moving average calculation of the recordSize with `serialized.remaining()` dynamically.

tzulitai replied:
A user-provided queue limit supplier function sounds like a good idea.
As you mentioned, this can come as a follow-up PR.

tzulitai replied:
The moving average calculation, that you described, could maybe just be a implementation of the limit supplier function.

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -55,6 +58,13 @@
+
+	public static final String METRIC_OUTSTANDING_RECORDS_COUNT = "outstandingRecordsCount";
+
+
```
nit: unnecessary line

fmthoma replied:
✔

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -144,6 +163,17 @@ public void setFailOnError(boolean failOnError) {
+	 * @param queueLimit The maximum length of the internal queue before backpressuring
+	 */
+	public void setQueueLimit(int queueLimit) {
+		this.queueLimit = queueLimit;
```
Will need argument checks on the given `queueLimit`.

fmthoma replied:
✔ (`queueLimit > 0`)

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +366,29 @@ private void checkAndPropagateAsyncError() throws Exception {
+			}
+			attempt++;
+			try {
+				backpressureLatch.await(100);
```
I like this implementation a lot better now 👍

tzulitai replied:
We might want to make the wait time configurable? (as a separate PR)
My reasoning is that it directly affects how long until the "flush taking unusually long" message starts popping up.

fmthoma replied:
It does, but if we make it configurable, I'd rather keep the warning threshold at one second, i.e. `if (attempt >= 1000 / waitTime) { … }`.

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/TimeoutLatch.java:
```diff
@@ -0,0 +1,41 @@
+
+package org.apache.flink.streaming.connectors.kinesis.util;
+
+public class TimeoutLatch {
```
This needs to be annotated as `@Internal`

fmthoma replied:
✔

tzulitai commented on flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducerTest.java:
```diff
@@ -267,6 +268,79 @@ public void go() throws Exception {
+		};
+		msg1.start();
+		msg1.trySync(100);
+		assertFalse("Flush triggered before reaching queue limit", msg1.isAlive());
```
I wonder if this would introduce flakiness in the test.
@fmthoma could you elaborate a bit here?

fmthoma replied:
@tzulitai In principle, yes, if the call `testHarness.processElement(…)` takes more than 100 milliseconds. However, I believe this is very unlikely even on slow systems, since the operation is mostly (entirely?) CPU bound. If test failures occur nevertheless, it should be no problem to increase the timeout for `msg1` and `msg2`.

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java:
```diff
@@ -326,6 +366,29 @@ private void checkAndPropagateAsyncError() throws Exception {
+	 * We don't want to flush _all_ records at this point since that would
+	 * break record aggregation.
+	 */
+	private void enforceQueueLimit() {
```
Would like to request one more slight change here:
Let this method return a boolean that indicates whether or not flushing occurred.

The caller of this method can then use the flag to decide whether or not `checkAndPropagateAsyncError` is required.

fmthoma replied:
✔

fmthoma commented:
@tzulitai Thanks for your last review comments! I addressed them, and rebased the branch against master.

tzulitai commented:
Thanks @fmthoma, will proceed to merge this ..

tzulitai commented:
@fmthoma I've merged this manually. Thanks for the contribution.
Could you close this PR?

fmthoma commented:
Merged manually by squashing: 7d034d4
